lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+ *
+ * @version 3429af9d91e2c9efe1af85757499e5a308755f5f
+ * @link    https://github.com/ampproject/amp-toolbox/blob/3429af9d91e2c9efe1af85757499e5a308755f5f/packages/optimizer/lib/transformers/PreloadHeroImage.js
+ *
+ * @package ampproject/optimizer
+ */
+final class PreloadHeroImage implements Transformer
+{
+
+    /**
+     * Class(es) to apply to a serverside-rendered image element.
+     *
+     * @var string
+     */
+    const SSR_IMAGE_CLASS = 'i-amphtml-fill-content i-amphtml-replaced-content';
+
+    /**
+     * List of attributes to copy onto an SSR'ed image.
+     *
+     * @var string[]
+     */
+    const ATTRIBUTES_TO_COPY = [
+        Attribute::ALT,
+        Attribute::ATTRIBUTION,
+        Attribute::OBJECT_FIT,
+        Attribute::OBJECT_POSITION,
+        Attribute::REFERRERPOLICY,
+        Attribute::SRC,
+        Attribute::SRCSET,
+        Attribute::SIZES,
+        Attribute::TITLE,
+    ];
+
+    /**
+     * Images smaller than 150px are considered tiny.
+     *
+     * @var int
+     */
+    const TINY_IMG_THRESHOLD = 150;
+
+    /**
+     * Maximum number of hero images defined via data-hero attribute.
+     *
+     * @var int
+     */
+    const DATA_HERO_MAX = 2;
+
+    /**
+     * Configuration store to use.
+     *
+     * @var TransformerConfiguration
+     */
+    private $configuration;
+
+    /**
+     * Instantiate a PreloadHeroImage object.
+     *
+     * @param TransformerConfiguration $configuration Configuration store to use.
+     */
+    public function __construct(TransformerConfiguration $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->configuration->get(PreloadHeroImageConfiguration::PRELOAD_HERO_IMAGE) === false) {
+            return;
+        }
+
+        $heroImages    = $this->findHeroImages($document);
+        $referenceNode = $document->viewport;
+
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > self::DATA_HERO_MAX) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = self::DATA_HERO_MAX;
+        }
+
+        $isAmpStory = Amp::isAmpStory($document);
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->generatePreload($heroImages[$index], $document, $errors, $referenceNode);
+            if (! $isAmpStory) {
+                // AMP Stories don't support SSR'd <amp-img> yet.
+                // See https://github.com/ampproject/amphtml/issues/29850.
+                $this->generateImg($heroImages[$index], $document);
+            }
```

Comment:
Is this true? See https://github.com/ampproject/amphtml/issues/29850#issuecomment-676830337

The playground is showing that the HTML is no longer invalid.

Reply:
Ah, seems like that's not up-to-date anymore. I'll adapt (and introduce a PR for the JS version as well).